taichi/ui/backends/vulkan/renderable.cpp

```diff
@@ -0,0 +1,301 @@
+          throw std::runtime_error(
+              "shape of per_vertex_color should be the same as vertices");
+        }
+        update_renderables_vertices_x64(
+            mapped_vbo, (float *)info.per_vertex_color.data, num_vertices, 3,
+            offsetof(Vertex, color));
+      }
+      if (info.normals.valid) {
+        if (info.normals.shape[0] != num_vertices) {
+          throw std::runtime_error(
+              "shape of normals should be the same as vertices");
+        }
+        update_renderables_vertices_x64(mapped_vbo, (float *)info.normals.data,
+                                        num_vertices, 3,
+                                        offsetof(Vertex, normal));
+      }
+      app_context_->device().unmap(staging_vertex_buffer_);
+
+      int *mapped_ibo =
+          (int *)app_context_->device().map(staging_index_buffer_);
+      if (info.indices.valid) {
+        indexed_ = true;
+        update_renderables_indices_x64(mapped_ibo, (int *)info.indices.data,
+                                       num_indices);
+      } else {
+        indexed_ = false;
+      }
+      app_context_->device().unmap(staging_index_buffer_);
+    }
+    app_context_->device().memcpy(vertex_buffer_.get_ptr(0),
+                                  staging_vertex_buffer_.get_ptr(0),
+                                  config_.vertices_count * sizeof(Vertex));
+    app_context_->device().memcpy(index_buffer_.get_ptr(0),
+                                  staging_index_buffer_.get_ptr(0),
+                                  config_.indices_count * sizeof(int));
+  } else {
+    throw std::runtime_error("unsupported field source");
+  }
+}
+
+Pipeline &Renderable::pipeline() {
+  return *(pipeline_.get());
+}
+
+const Pipeline &Renderable::pipeline() const {
+  return *(pipeline_.get());
+}
+
+void Renderable::create_bindings() {
+  ResourceBinder *binder = pipeline_->resource_binder();
+  binder->vertex_buffer(vertex_buffer_.get_ptr(0), 0);
+  binder->index_buffer(index_buffer_.get_ptr(0), 32);
+}
+
+void Renderable::create_graphics_pipeline() {
+  if (pipeline_.get()) {
+    return;
+  }
+  auto vert_code = read_file(config_.vertex_shader_path);
+  auto frag_code = read_file(config_.fragment_shader_path);
+
+  std::vector<PipelineSourceDesc> source(2);
+  source[0] = {PipelineSourceType::spirv_binary, frag_code.data(),
+               frag_code.size(), PipelineStageType::fragment};
+  source[1] = {PipelineSourceType::spirv_binary, vert_code.data(),
+               vert_code.size(), PipelineStageType::vertex};
+
+  RasterParams raster_params;
+  raster_params.prim_topology = config_.topology_type;
+  raster_params.depth_test = true;
+  raster_params.depth_write = true;
+
+  std::vector<VertexInputBinding> vertex_inputs = {{0, sizeof(Vertex), false}};
+  std::vector<VertexInputAttribute> vertex_attribs = {
+      {0, 0, BufferFormat::rgb32f, offsetof(Vertex, pos)},
+      {1, 0, BufferFormat::rgb32f, offsetof(Vertex, normal)},
```

Comment:
Can we potentially do RGB8 or RGBA8 for normals and colors? Should be doable since these buffers are coming from a CUDA kernel right

Reply:
Yep we can. Although do you think it's necessary to do that right now? Currently I'm using a single kernel to copy different vertex attributes: 

https://github.com/taichi-dev/taichi/blob/5d4796e0c852cd021502c11fc82d57a44c8a8e76/taichi/ui/backends/vulkan/renderables/kernels.h#L5-L14

Reply:
It would be less data to move around, you should be able to do that from within the same kernel.

Reply:
OK. And do we really wanna do this for normals? Does 8 bit normals have visible artifacts?

Reply:
should be enough at least for GGUI since our lighting don't rely on normals that much.

Reply:
we didn't define rgb8 in `device.h`...

Reply:
Use RGBA8, RGB8 is a 3 byte awkward format

Reply:
Let's leave this for now and add a TODO. I don't want to have to change device api and vulkan device in this PR..

Reply:
We should not change device API. RGB8 don't exist in Vulkan images as well since it's not aligned. Usual practice is just use RGBA8 and ignore the alpha channel

Reply:
![Screenshot from 2021-08-25 04-23-06](https://user-images.githubusercontent.com/16066115/130755717-e1e6889e-d085-42eb-b198-1ba435f5577d.png)
![Screenshot from 2021-08-25 04-26-15](https://user-images.githubusercontent.com/16066115/130755737-75e70522-b416-4cd4-aeff-b9a7e1521107.png)

Reply:
oops sorry for the dual screen.

Reply:
tried using 8 bit normals, looks really weird around the edges.

Reply:
gonna leave this for now :)

Reply:
Right by default it rgba8 means rgba8unorm, you will need the signed normalized version or do the range conversion (-128..127) to (0..255) manually

Reply:
K. will open a separate pr to add a `BufferFormat::rgba8signed` and fix this :)